dev/com.ibm.ws.microprofile.metrics.2.0.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetricsHandler.java

```diff
@@ -123,35 +126,70 @@ protected void unregister(String objectName) {
 	}
 
     protected void register() {
-		MBeanServer	mbs = ManagementFactory.getPlatformMBeanServer();
-		for (String sName : mappingTable.getKeys() ) {
-			Set<ObjectInstance> mBeanObjectInstanceSet;
-			try {
-				mBeanObjectInstanceSet = mbs.queryMBeans(new ObjectName(sName), null);
-				if (sName.contains("ThreadPoolStats")) {
-					final int MAX_TIME_OUT = 5000;
-					int currentTimeOut = 0;
-					while (mBeanObjectInstanceSet.isEmpty() && currentTimeOut <= MAX_TIME_OUT) {
-						Thread.sleep(50);
-						mBeanObjectInstanceSet = mbs.queryMBeans(new ObjectName(sName), null);
-						currentTimeOut+=50;
-					}
-				}
-		        for (ObjectInstance objInstance : mBeanObjectInstanceSet) {
-		            String objectName = objInstance.getObjectName().toString();
-		            String[][] data = mappingTable.getData(objectName);
-					if (data != null) {
-						register(objectName, data);
-		            }
-		        }
-			} catch (Exception e) {
-	            if (tc.isDebugEnabled()) {
-	                Tr.debug(tc, "register exception message: ", e.getMessage());
-	                FFDCFilter.processException(e, MonitorMetricsHandler.class.getSimpleName(), "register:Exception");
-	            }
-			}
-		}
-	}
+        MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+        ReentrantLock lock = new ReentrantLock();
+        for (String sName : mappingTable.getKeys()) {
+            Set<ObjectInstance> mBeanObjectInstanceSet;
+
+            try {
+                lock.lock();
+                try {
+                    /*
+                     * ThreadPoolStats is launched in a separate thread. It will invoke
+                     * queryMbeans() as well Lock just in case.
+                     */
+                    mBeanObjectInstanceSet = mbs.queryMBeans(new ObjectName(sName), null);
+                } finally {
+                    lock.unlock();
+                }
+
+                if (sName.contains("ThreadPoolStats") && mBeanObjectInstanceSet.isEmpty()) {
+                    ExecutorService execServ = Executors.newSingleThreadExecutor();
+                    execServ.execute(() -> {
+                        final int MAX_TIME_OUT = 5000;
+                        int currentTimeOut = 0;
+                        Set<ObjectInstance> mBeanObjectInstanceSetTemp = mBeanObjectInstanceSet;
+                        while (mBeanObjectInstanceSet.isEmpty() && currentTimeOut <= MAX_TIME_OUT) {
+                            try {
+                                Thread.sleep(50);
+                                lock.lock();
+                                try {
+                                    mBeanObjectInstanceSetTemp = mbs.queryMBeans(new ObjectName(sName), null);
+                                } finally {
+                                    lock.unlock();
+                                }
+                                currentTimeOut += 50;
```

Comment:
Is there a possibility of endless loop here?  When `mbs.queryMBeans()` throws an exception, it will jump to the catch block and skips the increment of `currentTimeOut` and go back to the beginning of the while loop.